### PR TITLE
certloader: Fix reload on K8s Secret/ConfigMap update

### LINKED ISSUE #13621: Hubble certloader stops picking up new certificates
## Bug report

**How to reproduce the issue**

1. Install Hubble Server/Relay with mTLS enabled and the helm-generated certificates, e.g. `helm install cilium ./cilium    --namespace kube-system --set hubble.listenAddress=":4244" --set hubble.relay.enabled=true`
2. Run `helm upgrade --reuse-values`. This will renew the certificates as expected (e.g. `level=info msg="Certificate authority updated" config=tls-server` message in cilium-agent)
3. Run `helm upgrade --reuse-values` again. **This change is ignored by certloader (bug)**, even thought the underlying secret mount has changed.

A similar issue occurs with `certgen` (Helm value `hubble.tls.auto.method="cronJob"`) as reported by @kAworu 

Root cause seems to be that the fsnotify watcher watches the symlink target, which will become invalid after the first change. See https://github.com/fsnotify/fsnotify/issues/199 - however watching the symlink itself is probably also not sufficient, as it is never touched.

### PR DESCRIPTION
**Pull request by gandro**
This change fixes the certloader Watch function to use the new fswatcher
package which supports the event sequence emitted when Kubernetes 
updates a volume mount for a ConfigMap or Secret.

The previous code did not work as expected on Kubernetes. Kubernetes
sets up a volume mount with symlinks containing a double indirection,
see e.g. spf13/viper#415 (comment)

Simply adding the certificate paths to fsnotify therefore is not
sufficient, as fsnotify watches the symlink target, which gets removed
once Kubernetes updates the volume. Therefore the certloader code would
handle a single update before its fsnotify watches became stale and it
would stop picking up new changes.

This change uses the newly introduced fswatcher package instead. It
wraps fsnotify, but is designed to work with non-existing paths and
symlink structures which contain a double indirection.

See the code comments for details:
https://github.com/cilium/cilium/blob/cda1f045b7489b058cb1fa00e741f8a70d060105/pkg/crypto/certloader/fswatcher/fswatcher.go#L36-L61

Fixes: #13621

### CODE REVIEW

**Comment by gandro** [comment]
For those who are curious, this is the sequence now reported when I use `helm upgrade` to regenerate the Hubble server certificates:

```
2020-10-19T20:38:50.148164825Z level=debug msg="Received fsnotify event" config=tls-server file-path=/var/lib/cilium/tls/hubble/..2020_10_19_20_38_50.999732679 operation=CREATE subsys=hubble
2020-10-19T20:38:50.148202400Z level=debug msg="Received fsnotify event" config=tls-server file-path=/var/lib/cilium/tls/hubble/..2020_10_19_20_38_50.999732679 operation=CHMOD subsys=hubble
2020-10-19T20:38:50.150342333Z level=debug msg="Received fsnotify event" config=tls-server file-path=/var/lib/cilium/tls/hubble/..data_tmp operation=CREATE subsys=hubble
2020-10-19T20:38:50.150367686Z level=debug msg="Received fsnotify event" config=tls-server file-path=/var/lib/cilium/tls/hubble/..data_tmp operation=RENAME subsys=hubble
2020-10-19T20:38:50.150376211Z level=debug msg="Received fsnotify event" config=tls-server file-path=/var/lib/cilium/tls/hubble/..data operation=CREATE subsys=hubble
2020-10-19T20:38:50.150382920Z level=debug msg="Received fsnotify event" config=tls-server file-path=/var/lib/cilium/tls/hubble/..2020_10_19_20_37_28.141813773 operation=REMOVE subsys=hubble
2020-10-19T20:38:50.249821117Z level=info msg="Keypair updated" config=tls-server keypair-sn=9ede50cfb87adc8b0de65d2f739744d1 subsys=hubble
2020-10-19T20:38:50.250465678Z level=info msg="Certificate authority updated" config=tls-server subsys=hubble
```

**Comment by gandro** [comment]
test-me-please

File: pkg/crypto/certloader/watcher.go [hunk 1]
```diff
@@ -118,61 +125,59 @@ func (w *Watcher) Watch() <-chan struct{} {
...
+				// events that we obtain within a certain time window and
+				// then reload _all_ certificates and keypairs at once.
+				// This drastically simplifies the implementation. As an added
+				// bonus, this approach also deals with the non-Kubernetes case
+				// where the keypair and ca-certificates files are updated
+				// non-atomically in individual writes (which this logic
+				// also coalesces).
+				if reload == nil {
+					reload = time.After(watcherCoalesceWindow)
 				}
```
Comment (aanm) [inline pkg/crypto/certloader/watcher.go]: am I reading this correctly? doesn't this mean that this is essentially polling?

Comment (gandro) [reply pkg/crypto/certloader/watcher.go]: Not really. The current approach suffers from spurious unrelated events causing unnecessary reloads, but it's still very different from polling. Note that the reload is only triggered when there is an event, not on a timer (which is what polling would do).

But I put the PR back into draft mode, there might be a more precise way to deal with the K8s sequence.

Comment (gandro) [reply pkg/crypto/certloader/watcher.go]: Resolved as outdated. The new approach does not suffer from spurious events anymore.

**Comment by gandro** [comment]
test-me-please

**Comment by gandro** [comment]
Removed @Rolinh from the requested reviewer, as he is unavailable right now. He has reviewed the overall approach via Zoom today but we have not done an in-depth code review.

File: pkg/crypto/certloader/fswatcher/fswatcher.go [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fswatcher
+
+import (
+	"errors"
+	"fmt"
+	"github.com/sirupsen/logrus"
```
Comment (christarazi) [inline pkg/crypto/certloader/fswatcher/fswatcher.go]: Nit: import ordering

File: pkg/crypto/certloader/fswatcher/fswatcher_test.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fswatcher
+
+import (
+	. "gopkg.in/check.v1"
```
Comment (christarazi) [inline pkg/crypto/certloader/fswatcher/fswatcher_test.go]: Nit: import ordering

File: pkg/crypto/certloader/fswatcher/fswatcher.go [hunk 4]
```diff
@@ -26,6 +25,7 @@ import (
 	"github.com/cilium/cilium/pkg/logging/logfields"
 
 	"github.com/fsnotify/fsnotify"
+	"github.com/sirupsen/logrus"
```
Comment (christarazi) [inline pkg/crypto/certloader/fswatcher/fswatcher.go:28]: Ah I see it was fixed here. Please squash this change with the previous commit

**Review by christarazi — CHANGES_REQUESTED** [review]
I didn't review the low-level logic and mostly stayed at a high-level. Just a few minor nits to clean up, otherwise the code looks good to me, and is nicely documented :smile:!

File: pkg/crypto/certloader/fswatcher/fswatcher.go (hunk 4 above)
Comment (gandro) [reply pkg/crypto/certloader/fswatcher/fswatcher.go:28]: Oh, good catch! Fixed, thanks!

**Comment by gandro** [comment]
test-me-please

File: pkg/crypto/certloader/fswatcher/fswatcher_test.go (hunk 3 above)
Comment (christarazi) [reply pkg/crypto/certloader/fswatcher/fswatcher_test.go]: You might have missed this one?

**Review by christarazi — APPROVED** [review]
LGTM, doesn't require another review from me on the last nit.

File: pkg/crypto/certloader/fswatcher/fswatcher_test.go (hunk 3 above)
Comment (gandro) [reply pkg/crypto/certloader/fswatcher/fswatcher_test.go]: Argh, yes! I must have messed up my rebase somehow (I did fix it, but then apparently commited the wrong file). Should be fixed now.

**Comment by gandro** [comment]
test-me-please

**Pull request merged**